conseil-common/src/main/scala/tech/cryptonomic/conseil/common/rpc/RpcClient.scala

```diff
@@ -0,0 +1,176 @@
+    with LazyLogging {
+
+  /**
+    * Method that converts [[fs2.Stream]] of JSON-RPC requests into [[fs2.Stream]]] of desired result objects,
+    * or throws the exception with the error. Request [[fs2.Stream]]] can only contain calls to one API method
+    * and can be batched to speed up the response time.
+    *
+    * @param batchSize The size of the batched request in single HTTP call
+    * @param request [[fs2.Stream]]] of the [[RpcRequest]]
+    *
+    * @return [[fs2.Stream]]] that contains result from JSON-RPC server call.
+    */
+  def stream[P, R](batchSize: Int)(request: Stream[F, RpcRequest[P]])(
+      implicit decode: EntityDecoder[F, Seq[RpcResponse[R]]],
+      encode: EntityEncoder[F, List[RpcRequest[P]]]
+  ): Stream[F, R] =
+    request
+      .chunkN(batchSize)
+      .evalTap(requests => debug(s"Call RPC in a batch of: ${requests.size}"))
+      // From JSON-RPC documentation: The Server MAY process a batch rpc call as a set of concurrent tasks, 
+      //                              processing them in any order and with any width of parallelism.
+      // So we also can process the responses in unordered order.
+      .mapAsyncUnordered(maxConcurrent) { chunks =>
+        httpClient
+          .expect[Seq[RpcResponse[R]]](
+            Method.POST(chunks.toList, Uri.unsafeFromString(endpoint), headers: _*)
+          )
+      }
+      .flatMap(Stream.emits)
+      .flatMap {
+        case RpcResponse(_, _, Some(error), _) =>
+          Stream.raiseError[F](error)
+        case RpcResponse(_, Some(result), _, _) =>
+          Stream(result)
+        case response =>
+          Stream.raiseError[F](
+            new UnsupportedOperationException(
+              s"Unexpected response from JSON-RPC server at $endpoint $response"
+            )
+          )
+      }
+
+  /**
+    * Helper method to log debug information.
+    * It wraps scalalogging call into [[Concurrent]] to prevent side effects.
+    * It should be moved to separate logging util in the future.
+    *
+    * @param message Debug message
+    */
+  private def debug(message: String): F[Unit] = Concurrent[F].delay(logger.debug(message))
+}
+
+object RpcClient {
+
+  /**
+    * Create [[cats.Resource]] with [[RpcClient]].
+    *
+    * @param endpoint JSON-RPC server url
+    * @param maxConcurrent Max concurrent processes used to fetch data from JSON-RPC server
+    * @param httpClient Http4s http client
+    * @param headers Additional http headers, e.g. authorization
+    */
+  def resource[F[_]: Concurrent](
+      endpoint: String,
+      maxConcurrent: Int,
+      httpClient: Client[F],
+      headers: Header*
+  ): Resource[F, RpcClient[F]] =
+    for {
+      client <- Resource.pure(
+        new RpcClient[F](
+          endpoint,
+          maxConcurrent,
+          httpClient,
+          headers: _*
+        )
+      )
+    } yield client
+
+  /**
+    * Exception with JSON-RPC error message
+    *
+    * Defined Error Codes:
```

Comment:
Just wonder if we couldn't write these error codes down in some sort of ADT or at least put them in form of hardcoded values like:
```
val NotWellFormedError = -32700
val UnsupportedEncodingError = -32701
```
Are we using them somewhere or they are here for the future?

Reply:
We don't use them, this is only the info for future use.